lib/actions/auth.js

```diff
@@ -0,0 +1,56 @@
+import { push } from 'connected-react-router'
+
+import { setPathBeforeSignIn } from '../actions/user'
+
+/**
+ * This function is called by the Auth0Provider component, with the described parameter(s),
+ * when a new access token could not be retrieved.
+ * @param {Error} err
+ * @param {AccessTokenRequestOptions} options
+ */
+export function showAccessTokenError (err, options) {
+  return function (dispatch, getState) {
+    // TODO: improve this.
+    console.error('Failed to retrieve access token: ', err)
+  }
+}
+
+/**
+ * This function is called by the Auth0Provider component, with the described parameter(s),
+ * when signing-in fails for some reason.
+ * @param {Error} err
+ */
+export function showLoginError (err) {
+  return function (dispatch, getState) {
+    // TODO: improve this.
+    if (err) dispatch(push('/oops'))
+  }
+}
+
+/**
+ * This function is called by the Auth0Provider component, with the described parameter(s),
+ * after the user signs in.
+ * @param {Error} err
+ */
+export function processSignIn (appState) {
+  return function (dispatch, getState) {
+    if (appState && appState.urlHash) {
+      // At this stage after login, Auth0 has already redirected to /signedin (Auth0-whitelisted)
+      // which shows the AfterLoginScreen.
+      //
+      // Here, we save the URL hash prior to login (contains a combination of itinerary search, stop/trip view, etc.),
+      // so that the AfterLoginScreen can redirect back there when logged-in user info is fetched.
+      // (For routing, it is easier to deal with the path without the hash sign.)
+      const urlHashWithoutHash = (appState.urlHash.split('#')[1] || '/')
```

Comment:
Is it possible for the split to yield an array of length = 1? I'm wondering if it's risky to try to access the second element (for fear of array out of index error). Is a better string operation to use here `appState.urlHash.replace('#', '')`?

Reply:
I agree it'd be good to avoid array index out of bounds exceptions, although, I think a brute force replacement of the hash character isn't nuanced enough.

Reply:
Using substring instead of array.split in bb9f9e8926a9e638be4f1a9a33340e6c48792067.